src/js/actions/libraries.js

```diff
@@ -140,59 +174,66 @@ define(function (require, exports) {
             return Promise.resolve();
         }
 
-        var currentLayer = currentLayers.first(),
+        var firstLayer = currentLayers.last(), // currentLayers are in reversed order
             IMAGE_ELEMENT_TYPE = "application/vnd.adobe.element.image+dcx",
             representationType = "image/vnd.adobe.photoshop", // This is the default type
             newElement;
 
         // However, if the layer is a smart object, and is owned by some other app, we need to change representation
         // we do this by matching extensions
-        if (currentLayer.isSmartObject()) {
-            var layerFileName = currentLayer.smartObject.fileReference,
-                fileExtension = pathUtil.extension(layerFileName);
+        if (currentLayers.size === 1 && firstLayer.isSmartObject()) {
+            var layerFileName = firstLayer.smartObject.fileReference;
+
+            // layerFileName will be undefined if CC Libraries is uploading the same layer.
+            if (!layerFileName) {
+                return Promise.resolve();
+            }
+
+            var fileExtension = pathUtil.extension(layerFileName);
 
             if (_EXTENSION_TO_REPRESENTATION_MAP.hasOwnProperty(fileExtension)) {
                 representationType = _EXTENSION_TO_REPRESENTATION_MAP[fileExtension];
             }
         }
 
-        var tempLayerName = (new Date().getTime()).toString();
-
-        return os.getTempFilename(tempLayerName)
-            .bind(this)
-            .then(function (tempFilename) {
+        return _getTempPaths().bind(this)
+            .then(function (paths) {
                 // Export the selected layers
 
-                var tempPath = path.dirname(tempFilename.path),
-                    tempPreviewPath = tempPath + "/preview.png",
-                    previewSize = { w: 248, h: 188 },
-                    exportObj = libraryAdapter.exportLayer(tempPath, tempPreviewPath, tempLayerName, previewSize);
+                var previewSize = { w: RENDITION_SIZE, h: RENDITION_SIZE },
+                    exportObj = libraryAdapter.exportLayer(paths.tempBasePath, paths.tempPreviewPath,
+                        paths.tempName, previewSize);
 
-                return descriptor.playObject(exportObj);
+                return descriptor.playObject(exportObj).then(function (saveData) {
+                    paths.exportedLayerPath = saveData.in._path;
+                    return paths;
```

Comment:
If you're just returning path and there is no asynchronicity, you can just put it in a var and get rid of this promise callback block.